core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -1156,6 +1156,62 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
    */
   final def tracingStatus(flag: TracingStatus): ZIO[R, E, A] = new ZIO.TracingStatus(self, flag)
 
+  /**
+   * Applies `f` if the predicate fails.
+   */
+  final def filterOrElse[R1 <: R, E1 >: E, A1 >: A](p: A => Boolean)(f: A => ZIO[R1, E1, A1]): ZIO[R1, E1, A1] =
+    self.flatMap {
+      case v if !p(v) => f(v)
+      case v          => new ZIO.Succeed(v)
+    }
+
+  /**
+   * Supplies `zio` if the predicate fails.
+   */
+  final def filterOrElse_[R1 <: R, E1 >: E, A1 >: A](p: A => Boolean)(zio: => ZIO[R1, E1, A1]): ZIO[R1, E1, A1] =
+    filterOrElse[R1, E1, A1](p)(_ => zio)
+
+  /**
+   * Fails with `e` if the predicate fails.
+   */
+  final def filterOrFail[E1 >: E](p: A => Boolean)(e: => E1): ZIO[R, E1, A] =
+    filterOrElse_[R, E1, A](p)(ZIO.fail(e))
+
+  /**
+   * Fail with `e` if the supplied `PartialFunction` does not match, otherwise
+   * succeed with the returned value.
+   */
+  final def collect[E1 >: E, B](e: E1)(pf: PartialFunction[A, B]): ZIO[R, E1, B] =
+    collectM(e)(pf.andThen(ZIO.succeed))
+
+  /**
+   * Fail with `e` if the supplied `PartialFunction` does not match, otherwise
+   * continue with the returned value.
+   */
+  final def collectM[R1 <: R, E1 >: E, B](e: E1)(pf: PartialFunction[A, ZIO[R1, E1, B]]): ZIO[R1, E1, B] =
+    self.flatMap {
+      case v if pf.isDefinedAt(v) => pf(v)
```

Comment:
Would it be possible to reformulate this (and `rejectM` below) using `PartialFunction#applyOrElse`? That would prevent double-evaluation of `pro.

Reply:
Ah, yes, `applyOrElse`! I knew there was something added to avoid double evaluation, but I completely forgot what it was.